File: src/main/java/leetcode/CombinationSum.java
package leetcode;


import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CombinationSum {


    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(candidates);
        combinationSumHelper(result, candidates, target, 0, new ArrayList<>());
        return result;
    }

    private void combinationSumHelper(List<List<Integer>> result, int[] candidates, int target, int start, List<Integer> sumList) {
        if (target == 0) {
            result.add(new ArrayList<>(sumList));
        }
        if (target < 0 || target < candidates[start]) {
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            sumList.add(candidates[i]);
            combinationSumHelper(result, candidates, target - candidates[i], i, sumList);
            sumList.remove(sumList.size() - 1);
        }
    }
}

File: golang/main/CombinationSum.go
package main

import "fmt"

func combinationSum(candidates []int, target int) [][]int {

	var a = [][]int{
		{0, 1, 2, 3},
		{4, 5, 6, 7},
		{8, 9, 10, 11},
	}
	return a
}

func main() {
	fmt.Print(combinationSum([]int{2, 3, 6, 7}, 7))
	fmt.Print("\n")
	fmt.Print(combinationSum([]int{2, 3, 5}, 8))
}
